### PR TITLE
switch to a more declarative config for bundle_conda

### PR DESCRIPTION
**Pull request by jaimergp**
# Description

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

# References

* https://github.com/napari/packaging/issues/14
* https://github.com/napari/napari/issues/1001#issuecomment-1090348981

# How has this been tested?

- [x] conda bundle CI passes

## Final checklist:
- [X] My PR is the minimum possible work for the desired functionality
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by jaimergp** [comment]
@goanpeca - this is ready for a first look! the main changes are in `setup.cfg`, the rest is to accommodate the implications of having the specs out of conda-native metadata files :D

**Review by Carreau — APPROVED** [review]
Far from being knowledgeable in this area, but all looks good to me.

**Comment by jni** [comment]
@jaimergp could you write a bit more about all the work here? Some notes:

- I don't love the "$channels" and "$dependencies" in the workflow — seems opaque, I don't know where they come from?
- For me, the whole point of using setup.cfg is to avoid duplicating the dependency declarations. If we are going to have a whole new section with the same dependencies (in addition to others), then why not just have a meta.yaml?

**Comment by jaimergp** [comment]
@jni 

* `$channels` and `$dependencies` come from [here](https://github.com/napari/napari/pull/4404/files#diff-afc0068c04c2f82b5ba745d52a251795a27940607f0257cb7861a9d687ab8909R46-R47), two lines above. That shell script takes the data from `setup.cfg` (notice how we pass the "key" of the section as an argument). Why shell? Because we need this data to set up the Python environment, so we don't have it around yet.

* I don't like it either because it's almost redundant, but it's not :/ Packages are mostly the same, but some have subtly different names, and we also need to deal with conda-specifics like variant selectors and so on. It also uses a slightly different syntax. Most version operators are the same, but some of the ones used in PyPI are not available in `conda` (like `=~`).
* We "have" a meta.yaml in the conda-forge feedstock already, which we are cloning in the CI, and then patching programmatically through `bundle_conda.py --patch-recipe` (new addition in this PR) to override what we have in the stable release with the latest needs of our development version. Before this we were blindly taking the latest release as source of truth, potentially producing bad nightlies. The problem here is that our conda-forge releases come _after_ the PyPI releases and the `constructor` installers.

I still need to update the docs to reflect our final workflow, but I am asking for this kind of feedback first to ensure we all like it.

File: bundle_conda.py [hunk 1]
```diff
@@ -119,7 +136,154 @@ def _generate_background_images(installer_type, outpath="resources"):
         clean_these_files.append(output)
 
 
-def _constructor(version=_version(), extra_specs=None):
+def _patch_napari_recipe(recipe_path: str):
+    from souschef.jinja_expression import set_global_jinja_var
```
Comment (goanpeca) [inline bundle_conda.py:172]: We could add a general docstring with the reason why we do this?

File: setup.cfg [hunk 2]
```diff
@@ -222,3 +222,123 @@ type = layers
...
+    tomlkit
+    ruamel.yaml
+    anaconda-client
+
+build_packages_channels =
+    conda-forge
+
+# run requirements in the napari conda package
+# this needs to be in sync with pypi's, a bit above
+napari_recipe_run =
```
Comment (goanpeca) [inline setup.cfg:251]: @jaimergp we could perhaps add a test that checks that this list (conda) and the one above (pip) are in sync.

This should help us keep things synced. 

I do see a lot of value in keeping all these pinnings in a single file (even if duplicate) than putting them in a separate `environment.yaml`

File: tools/setup_cfg_to_yaml.sh [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+set -euo pipefail
+
+setup_cfg_block_to_yaml_list () {
+    # 1. use awk in "paragraph mode" (RS= ) to get the block of contiguous lines
+    #    that contains the pattern passed in the command line; we use exit to quit
+    #    after printing the first result only
+    # 2. then get the lines that start with 1+ spaces (section indentation);
+    #    these are the dependencies we need to install with conda
+    # 3. then replace the spaces with the expected list marker for the yaml file
```
Comment (goanpeca) [inline tools/setup_cfg_to_yaml.sh:11]: ❤️

**Review by goanpeca — COMMENTED** [review]
Looks great @jaimergp!

I would just add a test to ensure the dependencies are in sync in the cfg file :) 

I did something like this a loooong time ago (see https://github.com/spyder-ide/spyder/blob/master/spyder/tests/test_dependencies_in_sync.py) to keep setup.py, requirements.txt and environment yam in sync.

(There is also the binder file, but we can take care of that in a separate round)

**Comment by jni** [comment]
> We "have" a meta.yaml in the conda-forge feedstock already, which we are cloning in the CI, and then patching programmatically through bundle_conda.py --patch-recipe (new addition in this PR) to override what we have in the stable release with the latest needs of our development version. Before this we were blindly taking the latest release as source of truth, potentially producing bad nightlies. The problem here is that our conda-forge releases come after the PyPI releases and the constructor installers.

Ok, but why don't we have a meta.yaml and/or [bundle-]environment.yml in *this* repo, keep it up to date with a test like @goanpeca proposes, and then come release we copy that over to the conda-forge recipe? Then we might not need shell scripts to convert cfg to yml?

**Comment by jaimergp** [comment]
Sure, we can do that as well. 

I was trying to centralize metadata instead of adding more files, but if we have tests to tie things together, that might be enough. That said, I am unsure how we can test for perfect equality when the restraints are expressed with slightly different syntax (maybe error out when an operator outside the common subset is used?). I need to look deeper into that code Gonzalo shared, but it looks like a way of solving it.

So, just to make sure I got this right:

* **CI environments**: Remove blocks from setup.cfg that require the shell snippet - these blocks will go into environment files directly.
* **Conda packages**: Not sure about this one: are you suggesting that we also create a meta.yaml with the definitions we want instead of patching the feedstock? We will need to make sure that our copy of meta.yaml is more or less in sync with conda-forge's, but that should be ok. We basically change which copy is the source of truth: ours instead of conda-forge's.
* **Constructor dependencies**: are these ok in setup.cfg or do you prefer we have a `construct.yaml` file as well? We need to add some fields dynamically depending on the type of build, so that's why we had it generated in the script first. This PR now adds the ability to define the dependencies in setup.cfg instead of using Python to write the dictionary directly.

In case we decide to add all these files, I'd suggest using `resources/packaging/conda` to store the metadata.

**Comment by jni** [comment]
Yup, my feeling is that the cost of setup.cfg -> yaml conversion script, together with the inconsistencies *right within* a setup.cfg, is too high a price to pay to have all the metadata in a single file (an admirable goal indeed! :joy:).

> CI environments: Remove blocks from setup.cfg that require the shell snippet - these blocks will go into environment files directly.

:+1: 

> are you suggesting that we also create a meta.yaml with the definitions we want instead of patching the feedstock?

I think so? :joy: I am not sure about this. Can you summarise our alternatives here? What do we do now? My impression is: on each release, we manually synchronise the values in the feedstock meta.yaml with the setup.cfg. In the meantime, nothing happens. But, this means that the nightlies are inconsistent — how are the nightlies built?

I'm also curious with the vision of how this will intersect with the idea of napari "distributions" for plugins, where we synchronise the packages that plugins will need to depend on for major packages like pytorch, tensorflow, cupy, etc. (Though I know that's out of scope for this PR :joy:)

> Constructor dependencies: are these ok in setup.cfg or do you prefer we have a construct.yaml file as well?

This depends: can the tooling install them from the setup.cfg or do you need a yaml? Again, I would prefer *not* to shoehorn them into a single file if they are going to be shuttled off into the required format by some custom scripting.

Overall, I understand the concern with the proliferation of files, but this can be mitigated by e.g. having them in a folder, and/or having some documentation around the whole thing — which is indeed now rather complex, but I see the necessity for everything. :sweat_smile:

**Comment by jaimergp** [comment]
> But, this means that the nightlies are inconsistent — how are the nightlies built?

Right now the workflow is the following:

1. Clone the conda-forge feedstock next to the napari repository. We end up with $GITHUB_WORKSPACE/napari-feedstock and $GITHUB_WORKSPACE/napari-source
2. Patch meta.yaml to make it use `$GITHUB_WORKSPACE/napari-source` instead of the latest release tarball. This is done with some shell/sed scripting. However, we were not doing anything about the runtime dependencies.
3. Run conda-smithy to rerender the feedstock so it catches the latest global configuration changes from conda-forge.

This PR changes step 2 in this way:

* Instead of using `sed` to replace strings, we use a Python script based on `souschef` (tool to edit conda recipes)
* We also patch the runtime dependencies with those stated in `setup.cfg`
* We add a 3rd package output, `napari-pinnings`,  which simply lists a set of `run_constrained` dependencies, as stated in `setup.cfg`. This is what we need to control the potential dependency issues in our plugin ecosystem.

I didn't want to publish `napari-pinnings` to conda-forge because it adds no value to that community and we don't want to re-release napari builds every time we patch the pinnings. Instead, we will only publish it to our own `napari` channel in anaconda.org. This is why I used this PR to centralize all the metadata, because I was going to use `souschef` anyway to add this new output package.

> This depends: can the tooling install them from the setup.cfg or do you need a yaml? Again, I would prefer not to shoehorn them into a single file if they are going to be shuttled off into the required format by some custom scripting.

It doesn't have native support for `setup.cfg`, but the massaging needed is equivalent to what `bundle.py` was doing. Each section is a block of strings, so we essentially parse that block of strings into separate lines, removing/keeping comments as necessary. We then use the resulting list to generate the `construct.yaml` file. This file however needs different fields across platforms (e.g. conda channels are different for arm64 due to the availability of pyqt/pyside packages).

***

I know this feels like a lot, because it is 😂  We are trying to standardize metadata usage across ecosystems with minimum maintenance, which will require some amount of either in-house scripting or reliance on external, experimental tooling. We need to find a compromise between "metadata-across-files hell" and "single-file with non-trivial scripting". The way I see it, it's feels easier to reason about problems when there are less moving pieces involved. Adding more files to the repo increases the amount of files I need to think about maintaining. If everything is in setup.cfg, I don't need to know that `conda build` wants a `meta.yaml`, `conda create` needs `environment.yml` and `constructor` requires `construct.yaml`.
This the reason I went for "all-things-in-setup.cfg": I find it simpler to think of a single file for metadata definitions if we ever run into packaging problems. 

I am not sure if developers will go and read the documentation when they find a problem or something they don't understand. Before they get to fix that dependency problem, they will have to understand all the files involved, and the different tools. (As a matter of fact, a part of this conversation is already covered in the current [docs](https://github.com/napari/napari/blob/main/docs/developers/packaging.md), so I wonder if they are serving the intended purpose).

I am happy to adjust the foundational principles of our metadata if that makes sense for everyone. I am comfortable with both a single-file approach and a multi-file one, but I'd like to know what other people, maybe less familiar with the conda ecosystem, think about this issue.

PS: We could also think of migrating from `setup.cfg` to `pyproject.toml` for better parsing... but that might be a can of worms. If editable installs are an issue, `flit` has support for it already, but `setuptools_scm` doesn't play nicely with it... aaa the wonders of packaging ✨ 📦

**Comment by jni** [comment]
> If everything is in setup.cfg, I don't need to know that conda build wants a meta.yaml, conda create needs environment.yml and constructor requires construct.yaml.
This the reason I went for "all-things-in-setup.cfg": I find it simpler to think of a single file for metadata definitions if we ever run into packaging problems.

I get this in general, but one thing that might/does happen is that you run the build and you get an error, and the error references a line in a file (e.g. meta.yaml) that doesn't exist, and then you can go crazy figuring out how it gets made and how that went wrong. I'm getting flashbacks to sphinx-gallery build failures for example.

Basically, the setup.cfg is, I think, a *very* [leaky abstraction](https://en.wikipedia.org/wiki/Leaky_abstraction).

Also, if I google "construct.yaml", I get taken to the constructor README, whereas if I google "setup.cfg conda_installer section" I will get nothing. (Though at some point I might get the right page on napari.org, yes ;)

But, it's definitely a tradeoff, I don't feel strongly enough to block this, but neither do I feel confident enough to approve. 😂 I agree with you, it'd be good to have more takes here. I'd particularly like to hear what @sofroniewn and @tlambert03 think.

For *either* solution, the *ideal* case for me is that we get the base dependencies from the pip dependencies, and update them in some declarative way that *might* include package renaming, e.g. a section like:

```
conda-package-names:
  - pyqt5: pyqt
conda-additional-deps:
  - python.app
```

I understand that this would be more engineering effort and I'm totally happy to punt on it, but I like to write down my ideals somewhere before we go down a compromise road. 😊

**Comment by sofroniewn** [comment]
> I'd particularly like to hear what @sofroniewn and @tlambert03 think.

I had a chat about this very generally with @nclack who seemed to favor the single file with custom build script approach in general under the rationale that it can be easier to maintain. Maybe he can give this a look over and weigh in on the PR too

**Comment by Czaki** [comment]
I suggest taking the approach from the minimal dependencies test and creating a python script that will be producing conda dependencies from python dependencies. As I understand there are only a few packages that need to be renamed so it could be handled with s small dict?

**Comment by jaimergp** [comment]
Thanks everyone for your inputs! There are a number of questions on the table we would need to address before deciding what to implement. Let me summarize:

1. Unified `setup.cfg` versus tool-specific files. If we don't like the supporting scripting (and that's fine), we will need the following files plus a test that ensures everything is in sync (not trivial, in my opinion, given the slightly different syntaxes and package names). Files:
  * Environments to provide the tooling needed in this CI: two different `environment.yml` files (although we can merge them in a single, larger environment if people prefer that).
  * Recipe: in principle, only the `meta.yaml` file, but we need to clone the full feedstock to have conda-forge compatible builds in our pipeline.
  * Installer: `construct.yaml`, which would need to be modified in some instances of the script to accommodate for platform-specific options. This is problematic because selectors in `conda`- world were implemented as comments (sigh), so round-trip parsing is lossy :/ That's why we build it programmatically now, inserting the packages and channels from setup.cfg.

2. What to do with the contents of `napari-pinnings`. We are patching conda-forge's `meta.yaml` to add a completely new output to the recipe. This might be too much for a "patch", depending on how we look at it. Its contents will also be key in the maintenance of the plugin ecosystem, so maybe it deserves a more prominent place. This might have a different release cadence than `napari` itself, so maybe we want it in a different package with its own recipe? Since this package will be used only internally in the installer, I don't think it should be posted to conda-forge, so we would keep it in our organization. Do we want to create a feedstock repo for it? 🤷

**Comment by jni** [comment]
> Since this package will be used only internally in the installer, I don't think it should be posted to conda-forge, so we would keep it in our organization.

I dunno, I kinda like the idea of having it in conda-forge so that pythonistas helping bundle users can create/troubleshoot in a proper conda-env of their own? Though I guess they can just add our channel in that situation?

**Comment by jaimergp** [comment]
> Though I guess they can just add our channel in that situation?

Yes, the idea is to post this to our `napari` channel only. The bundle installation will have `napari` on its channel lists, followed by `conda-forge`. To reproduce, they can do `conda create -n test -c napari -c conda-forge ...`.

**Comment by Czaki** [comment]
I would like to link this issue  napari/packaging#14, where I directly ask for such metapackage.

**Comment by tlambert03** [comment]
I feel rather unqualified to make much of a comment here.  I don’t have many strong opinions, and am  more or less happy to defer to whatever @jaimergp and @goanpeca think in this realm.  So with that caveat, my very weak thoughts here are:

- i don’t mind having additional files for metadata.  If there are standard places that these metadata should be (such as a constructor.yaml or something else), then my first inclination would be to put them there directly (even if it means one more root file).  If they _also_ supported setup.cfg, then great, but it does feel a bit weird to add it there just a place to consolidate.  (If anything, I guess I’d go for pyproject as the place that one day we might really consolidate everything). But very minor stuff
- i also agree it’d be lovely not to have to add runtime deps in two places, but see @jaimergp’s point that deps are slightly different in conda world and pip, so one way or the other, some special casing needs to be done.  

All told, this feels like a mostly internal decision that could be modified in the future, without breaking anything, right?

**Comment by jaimergp** [comment]
> All told, this feels like a mostly internal decision that could be modified in the future, without breaking anything, right?

Yes, this doesn't affect the end product. No matter how we organize the raw data, the building tools will end up receiving the same inputs, and hence will produce the same outputs.

I don't see `constructor` or `conda` supporting `pyproject.toml` configs any time soon, so I guess we'll go for more explicit files under `resources/packaging` or something like that. We might need to patch parts of them depending on the case, though, but it can be done.

**Comment by sofroniewn** [comment]
> Yes, this doesn't affect the end product. No matter how we organize the raw data, the building tools will end up receiving the same inputs, and hence will produce the same outputs.

Ok i think we can go ahead and merge this then. I think we've had a lot of folks weigh in and going with what has been implemented makes sense. If this causes us problems in the future we can change without effecting the end experience

@jaimergp can you fix the merge conflict - thanks!

**Review by sofroniewn — APPROVED** [review]

**Comment by jaimergp** [comment]
Conflicts are gone but that typing issue is still there.

**Comment by tlambert03** [comment]
Merge in main

**Review by nclack — APPROVED** [review]
lgtm

**Comment by sofroniewn** [comment]
Following https://napari.org/dev/naps/2-conda-based-packaging.html do we want to close this PR or switch over to draft? it's got a number of approves, so i don't want us to merge by mistake. @jaimergp what you recommend?

**Comment by jaimergp** [comment]
This will be taken over to `napari/packaging` as part of the ongoing migration, so no merge needed here! I marked it as draft just in case.

**Pull request closed without merging**